pkg/kubelet/pleg/generic.go

```diff
@@ -444,6 +444,11 @@ func updateRunningPodAndContainerMetrics(pods []*kubecontainer.Pod) {
 			}
 		}
 	}
+
+	// reset runningContainerCount to avoid stale metrics.
+	// For example, in last call, there's 1 container in unknown state, then the container's state changed to running
+	// state in current call, RunningContainerCount metrics should not keep 1 container in unknown state.
+	metrics.RunningContainerCount.Reset()
```

Comment:
The problem with doing it this way is that this basically introduces a race condition between populating all the other metric values for labels other than 'unknown'. If the scrape occurs between this line and the for loop before, you won't return any metrics. 

It would be better to remove the metrics with the label 'unknown' explicitly and ignore the other metrics. You can do this like so:

```
metrics.RunningContainerCount.Delete(map[string]string{"container_state": "unknown"})
```
or something to that effect.

Reply:
> It would be better to remove the metrics with the label 'unknown' explicitly and ignore the other metrics. You can do this like so. 
Yes,  all possible label values are  enumerated, and then  those not in current map will be removed.  I'm curious about how to make sure the metrics is right. Considering the situation that the scape occurs while the for loop is iterating, the GaugeVec has some values that is not updated, and some values updated.  Do we need to avoid such race condition? 

`for _, state := range []kubecontainer.State{
		kubecontainer.ContainerStateRunning,
		kubecontainer.ContainerStateCreated,
		kubecontainer.ContainerStateExited,
		kubecontainer.ContainerStateUnknown} {
		if _, ok := containerStateCount[string(state)]; !ok {
			metrics.RunningContainerCount.DeleteLabelValues(string(state))
		}
	}`

Reply:
No we wouldn't want to do that because there could be a race condition between the reset and set. In fact, what we really want to do is to check if "unknown" label occurs in our "set" for-loop and if that isn't true then after the for-loop where we set it, then we want to reset the "unknown" ones.

It's okay to have stale values in metrics. It's different to potentially return incorrect metrics, which the race condition basically allows.